### ISSUE TITLE
Silent NFQueue Fail

### ISSUE THREAD

**Issue by dhaavi** [safing/portmaster#110]
Creating a new issue for something that came up in another thread:

---

The network has stopped working in bar :(  `systemctl restart portmaster` fixed it as usually.

Sadly I have recently changed log level to info..

```
čec 25 23:01:37 northys-laptop portmaster-control[42451]: 200725 23:01:37.682 connection:207 ▶ INFO 076 filter: granting connection Unknown::-1 to maps.gstatic.com. (172.217.23.195), endpoint is not blacklisted (default=permit)
čec 25 23:01:37 northys-laptop portmaster-control[42451]: 200725 23:01:37.775 connection:207 ▶ INFO 077 filter: granting connection Unknown::-1 to maps.gstatic.com. (172.217.23.195), endpoint is not blacklisted (default=permit)
čec 25 23:01:37 northys-laptop portmaster-control[42451]: 200725 23:01:37.867 connection:207 ▶ INFO 078 filter: granting connection Unknown::-1 to maps.gstatic.com. (172.217.23.195), endpoint is not blacklisted (default=permit)
čec 25 23:01:49 northys-laptop portmaster-control[42451]: 200725 23:01:49.145 er/resolve:204 ▶ WARN 079 resolver: async query for www.gstatic.com.A failed: all 3 query-compliant resolvers failed, last error: read udp 192.168.101.52:58825->213.46.172.37:53: i/o timeout
čec 25 23:01:53 northys-laptop portmaster-control[42451]: 200725 23:01:53.983 tel/entity:170 ▶ WARN 080 intel: failed to resolve IP 192.168.101.42: failed to resolve 42.101.168.192.in-addr.arpa.PTR: all 1 query-compliant resolvers failed, last error: failed to send query: write udp6 [::]:44732->[ff02::fb]:5353: sendto: cannot assign requested address
čec 25 23:01:53 northys-laptop portmaster-control[42451]: 200725 23:01:53.983 connection:235 ▶ INFO 081 filter: dropping connection Unknown::-1 <- 192.168.101.42, endpoint is not whitelisted (incoming is always default=block)
čec 25 23:02:05 northys-laptop portmaster-control[42451]: 200725 23:02:05.556 tel/entity:170 ▶ WARN 082 intel: failed to resolve IP 192.168.101.1: failed to resolve 1.101.168.192.in-addr.arpa.PTR: all 1 query-compliant resolvers failed, last error: failed to send query: write udp6 [::]:44732->[ff02::fb]:5353: sendto: cannot assign requested address
čec 25 23:02:05 northys-laptop portmaster-control[42451]: 200725 23:02:05.556 connection:235 ▶ INFO 083 filter: dropping connection Unknown::-1 <- 192.168.101.1, endpoint is not whitelisted (incoming is always default=block)
čec 25 23:03:43 northys-laptop portmaster-control[42451]: 200725 23:03:43.552 tel/entity:170 ▶ WARN 084 intel: failed to resolve IP 192.168.101.11: failed to resolve 11.101.168.192.in-addr.arpa.PTR: all 1 query-compliant resolvers failed, last error: failed to send query: write udp6 [::]:44732->[ff02::fb]:5353: sendto: cannot assign requested address
čec 25 23:03:43 northys-laptop portmaster-control[42451]: 200725 23:03:43.552 connection:235 ▶ INFO 085 filter: dropping connection Unknown::-1 <- 192.168.101.11, endpoint is not whitelisted (incoming is always default=block)
čec 25 23:03:43 northys-laptop portmaster-control[42451]: 200725 23:03:43.556 all/master:081 ▶ INFO 086 filter: granting own connection root:/var/lib/portmaster/updates/linux_amd64/core/portmaster-core_v0-4-16:43322 <- 192.168.101.11
čec 25 23:03:43 northys-laptop portmaster-control[42451]: 200725 23:03:43.556 tel/entity:170 ▶ WARN 087 intel: failed to resolve IP fe80::149a:f07a:fe67:6b7c: failed to resolve c.7.b.6.7.6.e.f.a.7.0.f.a.9.4.1.0.0.0.0.0.0.0.0.0.0.0.0.0.8.e.f.ip6.arpa.PTR: all 1 query-compliant resolvers failed, last error: failed to send query: write udp6 [::]:44732->[ff02::fb]:5353: sendto: cannot assign requested address
čec 25 23:03:43 northys-laptop portmaster-control[42451]: 200725 23:03:43.556 connection:235 ▶ INFO 088 filter: dropping connection Unknown::-1 <- fe80::149a:f07a:fe67:6b7c, endpoint is not whitelisted (incoming is always default=block)





Welcome to fish, the friendly interactive shell
northys at northys-laptop in ~/c/b/backend (staging|✚1…3)
» dig root.cz
^C⏎ # did not work
northys at northys-laptop in ~/c/b/backend (staging|✚1…3)
» ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
^C
--- 8.8.8.8 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1026ms

northys at northys-laptop in ~/c/b/backend (staging|✚1…3)
» 
```

_Originally posted by @northys in https://github.com/safing/portmaster/issues/86#issuecomment-663906173_

---

I have to restart portmaster almost every wake up.

_Originally posted by @northys in a follow-up comment._

---

A few minutes ago I was logged to a server via SSH and I was able to execute commands but I couldn't start any new connection on my laptop (e.g. ping to 8.8.8.8.8 either without or with -n).

_Originally posted by @northys in a follow-up comment._

**Comment by dhaavi** [comment]
I think all of this has to do with the public network you are in. I would guess that there is a Captive Portal in the WiFi of that bar. If so, have you completed the Captive Portal process?

The Portmaster just lets `ping` through no matter what - without _any_ checks. So if ping does not work, there either is something terribly wrong with iptables and/ir the Integration with Portmaster - or the network you are connected to is responsible for that behavior.
Unless you have something fiddling around with iptables automatically, I think the latter is the case.

---

If you are experiencing issues like these, you can easily check for a Captive Portal like this:
- `curl -i neverssl.com`: The usual way that requires a server. See all whitelisted domains for connectivity checks [here](https://github.com/safing/portmaster/blob/04c846a0da3df4fc3fda02afdc366c102f447c54/netenv/online-status.go#L57-L100).
- `curl -i 255.255.255.254`: The Portmaster way of checking for a Captive Portal, _without_ leaking anything to a server. We are still experimenting with that, so please give us feedback!

What also could be happening, is that (1) your OS uses a domain for connectivity checks that is not in our list _and_ (2) the method we use to check for a Captive Portal does not work in that environment. In that case you potentially would not be notified of the Captive Portal and might conclude that something is broken.
If you use NetworkManager, you can find the configured domain used for connectivity checks somewhere in `/etc/NetworkManager/conf.d/*.conf`.

---

Also, please check if the Portmaster correctly identifies the local DNS server of the network. This does not depend on the network, but on the configuration of your OS. The logs show all configured and detected DNS servers on startup. It could look something like this:
```
200727 11:44:35.243 /resolvers:235 ▶ TRAC 062 resolver: loaded global resolvers:
200727 11:44:35.243 /resolvers:237 ▶ TRAC 063 resolver: dot://1.1.1.2:853?verify=cloudflare-dns.com&name=Cloudflare&blockedif=zeroip
200727 11:44:35.243 /resolvers:237 ▶ TRAC 064 resolver: dns://10.0.0.1:53
```
Line 064 shows the detected local resolver.

---

I also remember that you asked about how the Portmaster decides which DNS servers to use for a query. I'd love to answer that question, but it is going to take time. Fortunately, everything in that regard is nicely capsulated in a single file, which you can find [here](https://github.com/safing/portmaster/blob/develop/resolver/scopes.go).

**Comment by northys** [comment]
> I think all of this has to do with the public network you are in. I would guess that there is a Captive Portal in the WiFi of that bar. If so, have you completed the Captive Portal process?

The bar is the same ISP stock router I had when I lived in Prague. It has nothing special like captive portal or any other "business" stuff. Just "home" network and "guest" network with isolated subnets. Both encrypted with wpa2 perconal. I was connected to guest network when this happened. The only problem with this network is some packet loss time to time when someone is blocking signal by standing at the bar between me and the router.

The only Wi-Fi with captive portal I use regularly is the network in train - either local ČD or austrian OBB.

---

> What also could be happening, is that (1) your OS uses a domain for connectivity checks that is not in our list

I think it's just `connectivity-check.ubuntu.com` which is allowed.

I recall some DNS queries being blocked by nextdns (not using it anymore) which had female name in the hostname. Like data.anna.ubuntu.com or something. But I suppose it was just some stats collecting domain because the network worked even after those domains being blocked by nextdns (=resolving 0.0.0.0 as portmaster does when blocking stuff).

```
sudo journalctl -u portmaster.service --since "1 hour ago" | grep 'nameserver: handling new request' | grep ubuntu
čec 27 23:18:23 northys-laptop portmaster-control[169369]:             2.949µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.AAAA from 192.168.31.220:40661, getting connection
čec 27 23:18:23 northys-laptop portmaster-control[169369]:            16.322µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.AAAA from 127.0.0.1:35351, getting connection
čec 27 23:18:23 northys-laptop portmaster-control[169369]:              2.99µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.AAAA from 127.0.0.1:34702, getting connection
čec 27 23:19:54 northys-laptop portmaster-control[169369]:             3.145µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.A from 192.168.31.220:41613, getting connection
čec 27 23:23:23 northys-laptop portmaster-control[169369]:             2.366µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.AAAA from 192.168.31.220:34034, getting connection
čec 27 23:23:23 northys-laptop portmaster-control[169369]:             2.354µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.AAAA from 127.0.0.1:53339, getting connection
čec 27 23:23:23 northys-laptop portmaster-control[169369]:             2.219µs nameserver:175 ▶ TRAC     nameserver: handling new request for connectivity-check.ubuntu.com.AAAA from 127.0.0.1:37242, getting connection
```

> curl -i 255.255.255.254: The Portmaster way of checking for a Captive Portal, without leaking anything to a server. We are still experimenting with that, so please give us feedback!

I will!

---

>  Fortunately, everything in that regard is nicely capsulated in a single file, which you can find here.

Beer time... =)

**Comment by dhaavi** [comment]
Good news: I was on the right track with triaging.
Bad news: I was on the right track with triaging.

So, we finally succeeded in reproducing this error and it very much seems like an integration error with the Portmaster.

The easiest to prove the Portmaster is at fault is by doing localhost pings: `ping 127.0.0.1` and/or `ping6 ::1`. In our case, `ping 127.0.0.1` did not work, while `ping6 ::1` worked.

Somehow, one of the netfilter queues failed (silently), here we see only three running in a stack dump - there should be 4. In our case we lost either the IPv4 inbound or outbound queue, resulting in a total failure of IPv4 networking.
```
3 @ 0x40689b 0xa63e31 0xa652b9 0xa65300 0x468981
#	0xa63e30	github.com/safing/portmaster/firewall/interception/nfqueue._Cfunc_loop_for_packets+0x40	_cgo_gotypes.go:124
#	0xa652b8	github.com/safing/portmaster/firewall/interception/nfqueue.NewNFQueue.func1.1+0x58	/home/user/go/src/github.com/safing/portmaster/firewall/interception/nfqueue/nfqueue.go:63
#	0xa652ff	github.com/safing/portmaster/firewall/interception/nfqueue.NewNFQueue.func1+0x2f	/home/user/go/src/github.com/safing/portmaster/firewall/interception/nfqueue/nfqueue.go:63
```
To have the Portmaster output the stack to stderr (`.error.log` file), send it a `USR1` signal. You can do this by executing `sudo pkill -USR1 portmaster-core` (untested).

Now that we have found the culprit, we will start working on a fix. Don't know yet when it will be available, though. Thanks for hanging in there, @northys!

**Comment by ppacher** [comment]
Hi @northys,

we just released [**v0.4.18**](https://github.com/safing/portmaster/releases/tag/v0.4.18) that includes a new nfqueue backend that's expected to be much more stable and should fix this issue. Since we didn't have the time to thoroughly test it yet we decided to disable the new backend by default and put it behind a commandline flag (it works very well on our machines though).

It would be great if you could test the backend by enabling the `--experimental-nfqueue` option. Depending on how you installed the Portmaster in the first place you might want to run the latest installer again as it ships a new systemd service unit.

- Ubuntu/Debian: [Download from here](https://safing.io/portmaster/#downloads) and see [instructions here](https://github.com/safing/portmaster/wiki/Linux#ubuntu)
- Arch Linux: See [instructions here](https://github.com/safing/portmaster/wiki/Linux#arch-linux) (*we have a PKGBUILD now :tada:*)
- Custom: Please follow the [custom installation steps](https://github.com/safing/portmaster/wiki/Linux#custom-installation) again as we replaced `portmaster-control` with `portmaster-start`. *see #96 for more context on that switch*

Assuming the new `portmaster.service` file is in place, enable the new nfqueue backend by creating `/etc/default/portmaster` with the following content:
```
# You can adjust the log-level here now
LOGLEVEL=debug

PORTMASTER_ARGS="--experimental-nfqueue"
```

Finally, restart the Portmaster service by running

```
sudo systemctl daemon-reload
sudo systemctl restart portmaster
```

If anything doesn't work please feel free to ask!
We are looking forward to your feedback!

**Comment by northys** [comment]
@ppacher I wasn't able to update portmaster

sudo dpkg -i /home/northys/Downloads/portmaster-installer.deb 
(Reading database ... 258961 files and directories currently installed.)
Preparing to unpack .../portmaster-installer.deb ...
Unpacking portmaster (0.4.1.2) over (0.0.0-1) ...
Setting up portmaster (0.4.1.2) ...
Installing new version of config file /etc/xdg/autostart/portmaster_notifier.desktop ...
portmaster:  downloading modules, this may take a while
[control] 2020/08/02 21:53:47 WARNING: error loading indexes: failed to download index stable.json: failed to make request to "https://updates.safing.io/stable.json": Get "https://updates.safing.io/stable.json": dial tcp: lookup updates.safing.io: Temporary failure in name resolution
Error: failed to download index stable.json: failed to make request to "https://updates.safing.io/stable.json": Get "https://updates.safing.io/stable.json": dial tcp: lookup updates.safing.io: Temporary failure in name resolution
portmaster:  
portmaster:  WARN: Downloading modules failed!
portmaster:  downloading modules can be initiated by starting the Portmaster service or by running:
portmaster:  /var/lib/portmaster/portmaster-start --data=/var/lib/portmaster update
portmaster:  
portmaster:  installation successfull
Processing triggers for desktop-file-utils (0.24-1ubuntu3) ...
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...
Processing triggers for mime-support (3.64ubuntu1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...

# internet not working

» sudo reboot

# internet works

» sudo /var/lib/portmaster/portmaster-start --data=/var/lib/portmaster update
200802 23:58:38.084 ▶ BOF

» sudo systemctl restart portmaster
northys at northys-laptop in ~

» sudo journalctl -u portmaster.service --since "1 minute ago"
-- Logs begin at Tue 2020-07-28 19:44:45 CEST, end at Mon 2020-08-03 00:00:21 CEST. --
srp 02 23:59:30 northys-laptop systemd[1]: portmaster.service: Scheduled restart job, restart counter is at 5.
srp 02 23:59:30 northys-laptop systemd[1]: Stopped Portmaster by Safing.
srp 02 23:59:30 northys-laptop systemd[1]: Started Portmaster by Safing.
srp 02 23:59:30 northys-laptop systemd[6399]: portmaster.service: Failed to execute command: No such file or directory
srp 02 23:59:30 northys-laptop systemd[6399]: portmaster.service: Failed at step EXEC spawning /usr/bin/portmaster-control: No such file or directory
srp 02 23:59:30 northys-laptop systemd[1]: portmaster.service: Main process exited, code=exited, status=203/EXEC
srp 02 23:59:30 northys-laptop iptables[6411]: iptables: No chain/target/match by that name.
srp 02 23:59:30 northys-laptop iptables[6412]: iptables: No chain/target/match by that name.
srp 02 23:59:30 northys-laptop iptables[6413]: iptables: No chain/target/match by that name.
srp 02 23:59:30 northys-laptop systemd[1]: portmaster.service: Failed with result 'exit-code'.
srp 02 23:59:37 northys-laptop systemd[1]: Stopped Portmaster by Safing.


» ll | grep portmaster
-rwxr-xr-x 1 root root      12M čen 21 01:31 portmaster-control (deleted) # yup this is the name of the file including the (deleted)


# internet not working

northys at northys-laptop in /u/bin
» sudo /var/lib/portmaster/portmaster-start recover-iptables
Error: running [/usr/sbin/iptables -t mangle -C OUTPUT -j C170 --wait]: exit status 2: iptables v1.8.4 (legacy): Couldn't load target `C170':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.

# internet not working
```


REMOVED EVERYTHING, UNINSTALLED EVERYTHING AND INSTALLED PORTMASTER AGAIN AS SHOWN ABOVE


```
-- Logs begin at Tue 2020-07-28 19:44:45 CEST, end at Mon 2020-08-03 00:09:16 CEST. --
srp 03 00:08:32 northys-laptop systemd[1]: Started Portmaster by Safing.
srp 03 00:08:32 northys-laptop portmaster-start[4748]: [control] 2020/08/02 22:08:32 starting /var/lib/portmaster/updates/linux_amd64/core/portmaster-core_v0-4-18 --data /var/lib/portmaster --log info
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.196 ▶ BOF
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.196 ules/start:089 ▶ INFO 001 modules: initiating...
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.196 ules/start:186 ▶ INFO 002 modules: started database
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.196 ules/start:186 ▶ INFO 003 modules: started config
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.198 ules/start:186 ▶ INFO 004 modules: started rng
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.198 ules/start:186 ▶ INFO 005 modules: started base
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.198 ules/start:186 ▶ INFO 006 modules: started subsystems
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.198 ules/start:098 ▶ INFO 007 modules: initiated subsystems manager
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.298 dules/mgmt:070 ▶ INFO 008 modules: managing changes
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.298 ules/start:186 ▶ INFO 009 modules: started api
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.298 ules/start:186 ▶ INFO 010 modules: started netenv
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.298 ules/start:186 ▶ INFO 011 modules: started notifications
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.298 api/router:062 ▶ INFO 012 api: starting to listen on 127.0.0.1:817
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.305 ules/start:186 ▶ INFO 013 modules: started status
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.310 /resolvers:275 ▶ INFO 014 resolver: no scopes loaded
srp 03 00:08:32 northys-laptop portmaster-start[4748]: 200803 00:08:32.310 ules/start:186 ▶ INFO 015 modules: started resolver
srp 03 00:08:33 northys-laptop portmaster-start[4748]: 200803 00:08:33.300 ine-status:191 ▶ INFO 016 netenv: setting online status to SemiOnline (dns check query failed)
srp 03 00:08:35 northys-laptop portmaster-start[4748]: 200803 00:08:35.503 d-response:051 ▶ INFO 017 api request: 127.0.0.1:35582 HIJ /api/database/v1
srp 03 00:08:35 northys-laptop portmaster-start[4748]: 200803 00:08:35.503 i/database:084 ▶ INFO 018 api request: init websocket 127.0.0.1:35582 /api/database/v1
srp 03 00:08:36 northys-laptop portmaster-start[4748]: 200803 00:08:36.021 d-response:065 ▶ INFO 019 api request: 127.0.0.1:35584 307 /
srp 03 00:08:42 northys-laptop portmaster-start[4748]: 200803 00:08:42.305 dates/main:170 ▶ WARN 020 updates: failed to load indexes: failed to download index beta.json: failed to make request to "https://updates.safing.io/beta.json": Get "https://updates.safing.io/beta.json": dial tcp: lookup updates.safing.io: Te>
srp 03 00:08:42 northys-laptop portmaster-start[4748]: 200803 00:08:42.308 ules/start:186 ▶ INFO 021 modules: started updates
srp 03 00:08:42 northys-laptop portmaster-start[4748]: 200803 00:08:42.308 ules/start:186 ▶ INFO 022 modules: started geoip
srp 03 00:08:42 northys-laptop portmaster-start[4748]: 200803 00:08:42.309 ules/start:186 ▶ INFO 023 modules: started ui
srp 03 00:08:42 northys-laptop portmaster-start[4748]: 200803 00:08:42.319 ules/start:186 ▶ INFO 024 modules: started filterlists
srp 03 00:08:42 northys-laptop portmaster-start[4748]: 200803 00:08:42.319 ules/start:186 ▶ INFO 025 modules: started intel
srp 03 00:08:44 northys-laptop portmaster-start[4748]: [control] 2020/08/02 22:08:44 got terminated signal (ignoring), waiting for child to exit...
srp 03 00:08:44 northys-laptop portmaster-start[4748]:  <INTERRUPT>
srp 03 00:08:44 northys-laptop portmaster-start[4748]: 200803 00:08:44.328 e/run/main:075 ▶ WARN 026 main: program was interrupted, shutting down.
srp 03 00:08:44 northys-laptop systemd[1]: Stopping Portmaster by Safing...
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.304 d-response:051 ▶ INFO 027 api request: 127.0.0.1:35588 HIJ /api/database/v1
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.304 i/database:084 ▶ INFO 028 api request: init websocket 127.0.0.1:35588 /api/database/v1
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.650 d-response:065 ▶ INFO 029 api request: 127.0.0.1:35590 307 /
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.654 d-response:065 ▶ INFO 030 api request: 127.0.0.1:35590 200 /ui/modules/base/
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.667 d-response:065 ▶ INFO 031 api request: 127.0.0.1:35594 200 /ui/modules/base/js/chunk-vendors.d0291ca6.js
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.668 d-response:065 ▶ INFO 032 api request: 127.0.0.1:35590 200 /ui/modules/base/css/app.eff47ccc.css
srp 03 00:08:46 northys-laptop portmaster-start[4748]: 200803 00:08:46.668 d-response:065 ▶ INFO 033 api request: 127.0.0.1:35592 200 /ui/modules/base/js/app.ec6b5cdb.js
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.225 ater/fetch:074 ▶ INFO 034 updates: fetched https://updates.safing.io/all/ui/modules/assets_v0-1-5.zip (stored to /var/lib/portmaster/updates/all/ui/modules/assets_v0-1-5.zip)
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.227 d-response:065 ▶ INFO 035 api request: 127.0.0.1:35594 200 /assets/vendor/fonts/roboto.css
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.345 ater/fetch:074 ▶ INFO 036 updates: fetched https://updates.safing.io/all/intel/lists/index_v2020-3-25.dsd (stored to /var/lib/portmaster/updates/all/intel/lists/index_v2020-3-25.dsd)
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.352 ules/start:186 ▶ INFO 037 modules: started profiles
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.352 ules/start:186 ▶ INFO 038 modules: started processes
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.352 ules/start:186 ▶ INFO 039 modules: started network
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.357 d-response:051 ▶ INFO 040 api request: 127.0.0.1:35598 HIJ /api/database/v1
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.357 i/database:084 ▶ INFO 041 api request: init websocket 127.0.0.1:35598 /api/database/v1
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.383 d-response:065 ▶ INFO 042 api request: 127.0.0.1:35594 200 /assets/icons/level_normal.svg
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.383 d-response:065 ▶ INFO 043 api request: 127.0.0.1:35592 200 /assets/icons/level_high.svg
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.383 d-response:065 ▶ INFO 044 api request: 127.0.0.1:35590 200 /assets/icons/level_extreme.svg
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.423 d-response:065 ▶ INFO 045 api request: 127.0.0.1:35594 200 /assets/img/plants1-br.png
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.524 ules/start:186 ▶ INFO 046 modules: started interception
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.524 ules/start:186 ▶ INFO 047 modules: started core
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.524 ules/start:186 ▶ INFO 048 modules: started nameserver
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.524 ules/start:186 ▶ INFO 049 modules: started filter
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.524 dules/mgmt:088 ▶ INFO 050 modules: finished managing
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.524 dules/stop:043 ▶ WARN 051 modules: starting shutdown...
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.525 dules/stop:095 ▶ WARN 052 modules: could not stop module nameserver: dns: server not started
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.525 dules/stop:098 ▶ INFO 053 modules: stopped nameserver
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.526 dules/stop:098 ▶ INFO 054 modules: stopped filter
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.526 dules/stop:098 ▶ INFO 055 modules: stopped intel
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.526 dules/stop:098 ▶ INFO 056 modules: stopped geoip
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.526 dules/stop:098 ▶ INFO 057 modules: stopped filterlists
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.526 dules/stop:050 ▶ ERRO 058 modules: shutdown completed with error: modules: dependency loop detected, cannot continue
srp 03 00:08:47 northys-laptop portmaster-start[4748]: 200803 00:08:47.537 ◀ EOF
srp 03 00:08:47 northys-laptop portmaster-start[4748]: [control] 2020/08/02 22:08:47 core/portmaster-core exited without error
srp 03 00:08:47 northys-laptop iptables[5127]: iptables: No chain/target/match by that name.
srp 03 00:08:47 northys-laptop systemd[1]: portmaster.service: Succeeded.
srp 03 00:08:47 northys-laptop systemd[1]: Stopped Portmaster by Safing.
```

Systemd unit:
```
northys at northys-laptop in ~
» file /etc/default/portmaster
/etc/default/portmaster: cannot open `/etc/default/portmaster' (No such file or directory)
northys at northys-laptop in ~
» sudo systemctl cat portmaster.service
# /lib/systemd/system/portmaster.service
[Unit]
Description=Portmaster by Safing
Documentation=https://safing.io
Documentation=https://docs.safing.io
Before=nss-lookup.target network.target shutdown.target
After=systemd-networkd.service
Conflicts=shutdown.target
Wants=nss-lookup.target

# The Portmaster will conflict with almost all DNS services that
# that are listening on 53/upd or 53/tcp. Only DNS services that
# are likely to be running on desktop systems are listed here so
# systemd can gracefully stop them when portmaster is started.
# Note that manually starting one of the below services will instruct
# systemd to request portmaster to stop.
Conflicts=systemd-resolved.service dnsmasq.service unbound.service nextdns.service

[Service]
Type=simple
Restart=always
RestartSec=10
LockPersonality=yes
MemoryDenyWriteExecute=yes
NoNewPrivileges=yes
PrivateTmp=yes
PIDFile=/var/lib/portmaster/core-lock.pid
Environment=LOGLEVEL=info
Environment=PORTMASTER_ARGS=
EnvironmentFile=-/etc/default/portmaster
ProtectSystem=strict
ReadWritePaths=/var/lib/portmaster
# Portmaster calls iptables which requires to set xtables.lock
ReadWritePaths=/run/xtables.lock
RestrictAddressFamilies=AF_UNIX AF_NETLINK AF_INET AF_INET6
RestrictNamespaces=yes
RestrictSUIDSGID=yes
# In future version portmaster will require access to user home
# directories to verify application permissions.
ProtectHome=read-only
ProtectKernelTunables=yes
ProtectKernelLogs=yes
ProtectControlGroups=yes
PrivateDevices=yes
AmbientCapabilities=cap_chown cap_kill cap_net_admin cap_net_bind_service cap_net_broadcast cap_net_raw cap_sys_module cap_sys_ptrace cap_dac_override
CapabilityBoundingSet=cap_chown cap_kill cap_net_admin cap_net_bind_service cap_net_broadcast cap_net_raw cap_sys_module cap_sys_ptrace cap_dac_override
SystemCallArchitectures=native
#
```


Version (dashboard doesn't work...)
```
northys at northys-laptop in ~r.service^C
» sudo /var/lib/portmaster/portmaster-start version
Portmaster Start
version 0.4.1

commit tags/v0.4.17-0-g04c846a0da3df4fc3fda02afdc366c102f447c54
built with go1.14.2 (gc) linux/amd64
  using options 
  by user@docker
  on 24.07.2020

Licensed under the AGPLv3 license.
The source code is available here: https://github.com/safing/portmaster
```

**Comment by ppacher** [comment]
Hi @northys,

thanks for trying out and reporting back! I'm sad that it didn't work for you and I'm currently trying to reproduce this and will come back to you once I did!

In the meanwhile I can shed some lights on at least a few things:

<details>
<summary>`sudo dpkg -i /home/northys/Downloads/portmaster-installer.deb`</summary>

```bash

sudo dpkg -i /home/northys/Downloads/portmaster-installer.deb 
(Reading database ... 258961 files and directories currently installed.)
Preparing to unpack .../portmaster-installer.deb ...
Unpacking portmaster (0.4.1.2) over (0.0.0-1) ...
Setting up portmaster (0.4.1.2) ...
Installing new version of config file /etc/xdg/autostart/portmaster_notifier.desktop ...
portmaster:  downloading modules, this may take a while
[control] 2020/08/02 21:53:47 WARNING: error loading indexes: failed to download index stable.json: failed to make request to "https://updates.safing.io/stable.json": Get "https://updates.safing.io/stable.json": dial tcp: lookup updates.safing.io: Temporary failure in name resolution
Error: failed to download index stable.json: failed to make request to "https://updates.safing.io/stable.json": Get "https://updates.safing.io/stable.json": dial tcp: lookup updates.safing.io: Temporary failure in name resolution
portmaster:  
portmaster:  WARN: Downloading modules failed!
portmaster:  downloading modules can be initiated by starting the Portmaster service or by running:
portmaster:  /var/lib/portmaster/portmaster-start --data=/var/lib/portmaster update
portmaster:  
portmaster:  installation successfull
Processing triggers for desktop-file-utils (0.24-1ubuntu3) ...
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...
Processing triggers for mime-support (3.64ubuntu1) ...
Processing triggers for hicolor-icon-theme (0.17-2) ...

# internet not working

» sudo reboot
```
</details>

The new installer tries to download the required Portmaster modules right during installation. If that doesn't work it will attempt to download them later (or you can manually download them as you did with `portmaster-start update`). Seems like internet access wasn't working during installation time already. Since the systemd journal mentions `portmaster-control` not being found I guess portmaster was already running and died during the update because `portmaster-control` got removed (which may broke internet access as no DNS resolver was around anymore). 

It's curios that this behavior didn't pop up when I tried to upgrade a system with portmaster running. Nevertheless, I'll try again and hope to find and fix that issue soon. 

<details>
<summary>(deleted) files</summary>

```bash
» ll | grep portmaster
-rwxr-xr-x 1 root root      12M čen 21 01:31 portmaster-control (deleted) # yup this is the name of the file including the (deleted)
```
</details>
 
Thats fine, the linux kernel knows that the file has already been deleted but needs a reference for something. That's why it appends the `(deleted)` at the end of the filename. If you're curios you can checkout the related kernel code [here](https://elixir.bootlin.com/linux/latest/source/fs/d_path.c#L249)

<details>
<summary>portmaster-start recover-iptables</summary>

```bash
sudo /var/lib/portmaster/portmaster-start recover-iptables
Error: running [/usr/sbin/iptables -t mangle -C OUTPUT -j C170 --wait]: exit status 2: iptables v1.8.4 (legacy): Couldn't load target `C170':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
```
</details>

The error message is find as it just means that the portmaster did the cleanup of the C170 tables itself. Though, we should probably remove those error message to not confuse users.

IMHO internet wasn't working because no DNS service was available to serve lookup requests. We'll discuss a solution for that. In the meanwhile, if that happens again you may try to start the DNS service that's normally running on your system. `systemctl restart systemd-resolved` may do the trick (no sure which OS you're using).

<details>
<summary>/etc/default/portmaster</summary>

```
northys at northys-laptop in ~
» file /etc/default/portmaster
/etc/default/portmaster: cannot open `/etc/default/portmaster' (No such file or directory)
```
</details>

It's fine if that file does not exist. If it exists, it's read by the systemd unit but it is ignored if the file is missing. `EnvironmentFile=-/etc/default/portmaster` (the `-` prefix in the file-path tells systemd to ignore it in case of an error)

---

A last question, did internet work after removing and reinstalling everything? Or is the Portmaster still broken on your side?

**Comment by northys** [comment]
Hi @ppacher, thanks for fast reply!

> The new installer tries to download the required Portmaster modules right during installation. If that doesn't work it will attempt to download them later (or you can manually download them as you did with portmaster-start update). Seems like internet access wasn't working during installation time already. Since the systemd journal mentions portmaster-control not being found I guess portmaster was already running and died during the update because portmaster-control got removed (which may broke internet access as no DNS resolver was around anymore).

There was already new service which uses the new portmaster-start service. That's why I have posted it here so you could check if I have the new `portmaster.service` instaled from the downloaded `deb` package. **I expected it to be working after running portmaster update manually (as suggested by package in dpkg output) and restarting portmaster.service OR/AND rebooting my laptop**.

> IMHO internet wasn't working because no DNS service was available to serve lookup requests. We'll discuss a solution for that. In the meanwhile, if that happens again you may try to start the DNS service that's normally running on your system. systemctl restart systemd-resolved may do the trick (no sure which OS you're using).

That's why I rebooted ubuntu on my laptop, to "fix the internet". I didn't try to ping `8.8.8.8` though. Btw for some reason I remember the `update.safing.at` is CNAME to `something.safing.network` so I recall I digged it yesterday but do not remember in what context. I just want to put more mess into this :D I would say that the DNS was working (local cache from god knows where) and all but localhost didn't work. Just rumors and theories...

> The error message is fine as it just means that the portmaster did the cleanup of the C170 tables itself. Though, we should probably remove those error message to not confuse users.

I will create a feature request ticket.

> A last question, did internet work after removing and reinstalling everything? Or is the Portmaster still broken on your side?

I wouldn't be so calm while having broken networking on my laptop :rofl:

The portmaster just don't start because of missing plugins. I actually started the dashboard successfully but it losses connection immediately after start with message "no modules loaded".

**Comment by northys** [comment]
I've set the experimental Q flag.

**Comment by ppacher** [comment]
Thanks for trying this out, it should add more stability to Portmaster!

I hope the total loss of networking does not happen anymore and I'm looking forward to your answer! Note that this fix will likely not resolve #87 as this seems to be a different problem.

**Comment by dhaavi** [comment]
Hey @northys, it seems this issue has been resolved. I'm not sure how much you were able to test because of the other current issue, so please reopen this if something regarding this issue or change comes up again.
